remote_attestation/rust/src/crypto.rs

```diff
@@ -148,24 +155,54 @@ impl KeyNegotiator {
     }
 
     pub fn derive_session_key(self, peer_public_key: &[u8]) -> anyhow::Result<Vec<u8>> {
-        let peer_public_key =
-            agreement::UnparsedPublicKey::new(KEY_AGREEMENT_ALGORITHM, peer_public_key);
+        let public_key = self.public_key()?;
         let session_key = agreement::agree_ephemeral(
             self.private_key,
-            &peer_public_key,
+            &agreement::UnparsedPublicKey::new(KEY_AGREEMENT_ALGORITHM, peer_public_key),
             ring::error::Unspecified,
-            Self::key_derivation_function,
+            |key_material| {
+                Self::key_derivation_function(
+                    key_material,
+                    &public_key,
+                    &peer_public_key.as_ref().to_vec(),
+                )
+            },
         )
-        .map_err(|error| anyhow!("Couldn't derive session key: {:?}", error))?;
+        .map_err(|error| anyhow!("Couldn't agree on a session key: {:?}", error))?;
         Ok(session_key)
     }
 
-    /// Derives a session key from `key_material`.
-    fn key_derivation_function(key_material: &[u8]) -> Result<Vec<u8>, ring::error::Unspecified> {
-        // TODO(#2100): Derive a session key not just from `key_material` but also from server's and
-        // client's public keys.
-        // https://datatracker.ietf.org/doc/html/rfc7748#section-6.1
-        Ok(key_material.to_vec())
+    /// Derives a session key from `key_material` using HKDF.
+    /// https://datatracker.ietf.org/doc/html/rfc5869
+    /// https://datatracker.ietf.org/doc/html/rfc7748#section-6.1
+    ///
+    /// TODO(#2181): Use separate keys for server and client encryption.
+    fn key_derivation_function(
+        key_material: &[u8],
+        public_key: &[u8],
+        peer_public_key: &[u8],
+    ) -> Result<Vec<u8>, ring::error::Unspecified> {
+        // Session key is derived from a purpose string, public key and peer public key.
+        let mut info = vec![];
+        info.push(SESSION_KEY_PURPOSE.as_bytes());
+        // Sort public keys so that keys derived on the both sides of the protocol are equal.
+        if public_key.iter().cmp(peer_public_key.iter()) == Ordering::Less {
+            info.push(public_key);
+            info.push(peer_public_key);
+        } else {
+            info.push(peer_public_key);
+            info.push(public_key);
+        }
```

Comment:
Correct, it's not what is done in SC. They have separate keys for client and server.
It will be updated once https://github.com/project-oak/oak/issues/2181 is finished and we will have separate keys as well.